Add self-update command

amansilla opened the pull request:
This is a proposal for the self-update command #13 to be used as base although it needs probably some modifications.

Since the latest ```deptrac.phar``` is hosted in ```http://get.sensiolabs.de/deptrac.phar```, I thought the following location can be used to indicate its version ```http://get.sensiolabs.de/deptrac.version``` (currently not existing and should be created by SensionLabs as they admin the continuous delivery process I guess), which will be used in the code to fetch the latest version and compare it with the running one.

@slde-flash Let me know, what do you think about this or if you need a more sophisticated solution :smiley:.

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+    {
+        $this
+            ->setName('self-update')
+            ->setAliases(array('selfupdate'))
```
i don't think we need the alias

xabbuh replied:
At least, having this alias would be inline what Composer and the Symfony Installer offer (thus people might be used to use `selfupdate` instead of `self-update`).

slde-flash replied:
well, ok 👍

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+    public function execute(InputInterface $input, OutputInterface $output)
+    {
+        $latest = trim(file_get_contents('http://get.sensiolabs.de/deptrac.version'));
+        if (empty($latest)) {
```
nice idea but for now i don't want to maintain this file. can you just remove the "latest" check?

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+            return 1;
+        }
+
+        if ($this->getApplication()->getVersion() === $latest) {
```
no need for this check

amansilla replied:
Do you think we should update the phar always although there are no changes between the remote and the local version? I think would be nice check that e.g. using sha1, what do you think?

slde-flash replied:
i think we should keep it as simple as possible. from my point of view for the user it's just important, that he has the newest version.

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        $localFilename = realpath($_SERVER['argv'][0]) ?: $_SERVER['argv'][0];
+        $tmpDir = is_writable(dirname($localFilename)) ? dirname($localFilename) : false;
```
i think we can just use `sys_get_temp_dir` (no need to declare `$tmpDir`)

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        $tmpFilename = $tmpDir . '/deptrac-tmp.phar';
+        file_put_contents($tmpFilename, file_get_contents('http://get.sensiolabs.de/deptrac.phar'));
```
``` 
if(!($lastestDeptrac = file_get_contents('http://get.sensiolabs.de/deptrac.phar'))) {
   $output->writeln('<error>Could not download http://get.sensiolabs.de/deptrac.phar.</error>');
}
```

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+
+        $tmpFilename = $tmpDir . '/deptrac-tmp.phar';
+        file_put_contents($tmpFilename, file_get_contents('http://get.sensiolabs.de/deptrac.phar'));
+        if (false === file_exists($tmpFilename)) {
```
please just use `if (!file_exists($tmpFilename))` (no yoda conditions)

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+            return 1;
+        }
+
+        rename($tmpFilename, $localFilename);
```
we need a check if `$localFilename` is writeable

slde-flash commented:
thank you soo much, i added a bunch of small comments.
could you also update the docs for the new command?

slde-flash commented:
@amansilla i would like to have 100% coverage for the project. please let me know if you also want to work on the tests. i didn't tried it but i think it could be possible to use symfony's filesystem component in order to avoid mocking the filesytem.

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,57 @@
+        rename($tmpFilename, $localFilename);
+        $output->writeln('<info>Deprac was successfully updated.</info>');
+    }
+}
```
we should also make the phar file executable (chmod +x)

amansilla replied:
Do you think the phar file permissions being executed (so it already has execution permission) could be changed by the rename somehow? Otherwise I can't understand why that is necessary, just curiosity :wink:.

slde-flash replied:
you're right, no need for chmod +x

amansilla replied:
Maybe I wasn't right :joy:.

After a try, I experienced that new/renamed files in linux system are never going to have execution right independently of the umask set for the directory. Adding +x for all users would be insecure so I'd propose to read the original file permissions before the rename for setting it later to the updated file. That way the original file permissions stay untouched.

amansilla commented:
@slde-flash I've updated the command with your suggestions and added the documentation, so now it should be good enough. 

About the tests, I've being already thinking about it but I guess current the code is not fully testable without a previous refactoring. I guess we could use the filesystem component for the file handling and maybe guzzle for fetching the remote phar.

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,49 @@
+            $output->writeln('<error>Could not write deptrac.phar</error>');
+
+            return 1;
+        } else {
```
you can remove the else statement here

slde-flash commented on src/Command/SelfUpdateCommand.php:
```diff
@@ -0,0 +1,49 @@
+
+            return 1;
+        } else {
+            chmod($localFilename, $localFilePerm);
```
lets add a `@`, i dont think all filesystems supports this.

slde-flash commented:
i am 👍 using guzzle and the filesystem component.

WouterJ commented:
What about using https://github.com/padraic/phar-updater here?

amansilla commented:
I think it's a good option and simplifies the command, they only problem I see, this requires to maintain a new `deptrac.version` file in order to track the version. As I know, @slde-flash wanted to keep this as simple as possible.

slde-flash commented on composer.json:
```diff
@@ -3,18 +3,19 @@
     "license": "MIT",
     "require": {
         "php": ">=5.5.9",
+        "graphp/graphviz": "0.2.*",
```
can you try to not reorder the dependencies?

amansilla replied:
For sure, it'll be fixed on next update. I just like and am used to order the dependencies, I find it more readeable.

slde-flash replied:
feel free to create a PR that just reorders the dependencies =)

amansilla replied:
Ok, now I got your point of view. ;)

slde-flash commented on src/Tests/Command/SelfUpdateCommandTest.php:
```diff
@@ -0,0 +1,60 @@
+        $tmpFilename = sys_get_temp_dir() . '/deptrac.phar';
+
+        $this->httpClient->expects($this->once())
+            ->method('get')
```
all the tests are written using prophecy

slde-flash commented:
@amansilla would you prefer moving to https://github.com/padraic/phar-updater if i would maintain the deptrac.version file?

amansilla commented:
@slde-flash I'll to refactor this again, for using the phar updater as it seems to be easy to use and it's integration will simplify the testing as well. In case I find any new problem with it, I'll let you know.

slde-flash commented:
@amansilla thank you sooo much :)

amansilla commented:
@slde-flash I refactored the code for using the phar updater. Before I try to add some tests, could you please review the code?

slde-flash commented on src/Updater/DeptracStrategy.php:
```diff
@@ -0,0 +1,18 @@
+
+class DeptracStrategy extends ShaStrategy
+{
+    const PHAR_URL = 'http://get.sensiolabs.de/deptrac.phar';
```
please dont use CONST (because const is "public"), i am fine with inlining this or private static :)

amansilla replied:
The URL was inline added for both ;).

slde-flash commented:
@amansilla looks awesome. i am going to play with this pr tonight, if something is missing, i'll take over :)
if you want you could work on a PR that adds Unittests (using prophecy) to this code.
